File: core/r/R/cox_forest.R
# Recipe: Cox proportional-hazards model with a forest plot of hazard ratios.
# Tidy input: time, event (1 = event, 0 = censored), and covariate columns.
# The forest plot shows each covariate's hazard ratio with its 95% CI and
# p-value; the proportional-hazards assumption is checked and reported.

recipe_cox_forest <- function(df, spec) {
  time_col  <- spec$data$time
  event_col <- spec$data$event
  covs      <- spec$data$covariates
  if (is.null(time_col) || is.null(event_col)) {
    stop("cox_forest needs --time and --event columns")
  }
  if (is.null(covs) || length(covs) == 0) {
    covs <- setdiff(names(df), c(time_col, event_col))   # default: everything else
  }
  for (c in c(time_col, event_col, covs)) {
    if (!c %in% names(df)) stop(sprintf("column '%s' not found", c))
  }

  df[[time_col]] <- suppressWarnings(as.numeric(df[[time_col]]))
  ev <- suppressWarnings(as.integer(df[[event_col]]))
  if (!all(ev %in% c(0L, 1L) | is.na(ev))) {
    stop(sprintf("event column '%s' must be coded 1 = event, 0 = censored", event_col))
  }
  df[[event_col]] <- ev
  for (c in covs) if (is.character(df[[c]])) df[[c]] <- factor(df[[c]])

  n0   <- nrow(df)
  keep <- stats::complete.cases(df[, c(time_col, event_col, covs)]) &
          is.finite(df[[time_col]]) & df[[time_col]] >= 0
  df <- df[keep, , drop = FALSE]
  clean_steps <- if (nrow(df) < n0) {
    sprintf("Dropped %d row(s) with missing/invalid values (%d -> %d rows).",
            n0 - nrow(df), n0, nrow(df))
  } else "No cleaning applied; input used as-is."

  Surv <- survival::Surv
  rhs <- paste(sprintf("`%s`", covs), collapse = " + ")
  fml <- stats::as.formula(sprintf("Surv(`%s`, `%s`) ~ %s", time_col, event_col, rhs))
  cox <- survival::coxph(fml, data = df)

  zph <- tryCatch(survival::cox.zph(cox), error = function(e) NULL)
  ph_p <- if (!is.null(zph)) unname(zph$table["GLOBAL", "p"]) else NA_real_

  plot <- survminer::ggforest(cox, data = df)

  tid <- broom::tidy(cox, exponentiate = TRUE, conf.int = TRUE)
  stats_df <- data.frame(
    term    = tid$term,
    HR      = tid$estimate,
    ci_low  = tid$conf.low,
    ci_high = tid$conf.high,
    z       = tid$statistic,
    p_value = tid$p.value,
    stringsAsFactors = FALSE
  )

  test_meta <- list(
    name = "Cox proportional-hazards model",
    n = nrow(df), events = sum(df[[event_col]]),
    covariates = as.list(covs),
    proportional_hazards = list(test = "cox.zph global", p_value = ph_p),
    hazard_ratios = lapply(seq_len(nrow(stats_df)), function(i) list(
      term = stats_df$term[i], HR = stats_df$HR[i],
      ci_low = stats_df$ci_low[i], ci_high = stats_df$ci_high[i],
      p_value = stats_df$p_value[i]
    ))
  )

  resolved <- list(family = "cox", label = "Cox proportional hazards", covariates = covs)
  methods  <- .cox_methods(stats_df, ph_p, covs, nrow(df), sum(df[[event_col]]))
  build_script <- function(in_name, fig_stub) {
    .cox_emit_script(spec, time_col, event_col, covs, in_name, fig_stub)
  }

  n_display <- length(covs) + nrow(stats_df) + 3
  list(plot = plot, stats = stats_df, test_meta = test_meta, resolved = resolved,
       methods = methods, build_script = build_script, label = "hazard_ratios",
       df_used = df, clean_steps = clean_steps,
       width = 7.4, height = max(3.8, 0.42 * n_display + 1.2))
}

.cox_methods <- function(stats_df, ph_p, covs, n, events) {
  v    <- pkg_versions(c("survival", "survminer", "broom"))
  rver <- paste(R.version$major, R.version$minor, sep = ".")
  pfmt <- function(p) if (is.na(p)) "NA" else if (p < 0.001) "< 0.001" else formatC(p, format = "g", digits = 2)

  sig <- stats_df[which(stats_df$p_value < 0.05), , drop = FALSE]
  sig_txt <- if (nrow(sig) > 0) {
    paste0(" Significant predictors: ",
           paste(sprintf("%s (HR = %.2f, 95%% CI %.2f-%.2f, P = %s)",
                         sig$term, sig$HR, sig$ci_low, sig$ci_high,
                         vapply(sig$p_value, pfmt, character(1))), collapse = "; "),
           ".")
  } else " No individual predictor reached significance."
  ph_txt <- if (is.na(ph_p)) "" else {
    sprintf(" The proportional-hazards assumption was %s (global cox.zph P = %s).",
            if (ph_p < 0.05) "violated" else "not violated", pfmt(ph_p))
  }

  paste0(
    sprintf("Hazard ratios were estimated with a Cox proportional-hazards model (%d subjects, %d events) including %s.",
            n, events, paste(covs, collapse = ", ")),
    sig_txt, ph_txt,
    sprintf(" Analyses were performed in R %s with survival %s, and the forest plot was drawn with survminer %s (tidied with broom %s).",
            rver, v$survival, v$survminer, v$broom)
  )
}

.cox_emit_script <- function(spec, time_col, event_col, covs, in_name, fig_stub) {
  rhs <- paste(sprintf("`%s`", covs), collapse = " + ")
  c(
    "#!/usr/bin/env Rscript",
    "# Standalone reproduction. Run inside the pinned pubready container",
    "# (see REPRODUCE.md) from this bundle folder.",
    "suppressPackageStartupMessages({",
    "  library(survival); library(survminer); library(readr)",
    "})",
    "",
    sprintf('df <- as.data.frame(read_csv("%s", show_col_types = FALSE))', in_name),
    paste0(sprintf('for (c in c(%s)) if (is.character(df[[c]])) df[[c]] <- factor(df[[c]])',
                   paste(sprintf('"%s"', covs), collapse = ", "))),
    "",
    sprintf('cox <- coxph(Surv(`%s`, `%s`) ~ %s, data = df)', time_col, event_col, rhs),
    "p <- ggforest(cox, data = df)",
    "",
    sprintf('ggsave("%s.pdf", p, width = 7.4, height = 5.0)', fig_stub),
    'cat("Reproduced figure.\\n")'
  )
}

File: core/python/pubready/recipes/cox_forest.py
"""Recipe: Cox proportional-hazards model with a forest plot of hazard ratios
(lifelines). Mirrors the R cox_forest: HR with 95% CI and p per covariate, and a
proportional-hazards check."""
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from lifelines import CoxPHFitter

from ..theme import apply_pub_style
from ..util import format_p


def recipe_cox_forest(df, spec):
    t, e = spec["data"]["time"], spec["data"]["event"]
    covs = spec["data"].get("covariates")
    if not t or not e:
        raise ValueError("cox_forest needs --time and --event columns")
    if not covs:
        covs = [c for c in df.columns if c not in (t, e)]
    for c in [t, e] + covs:
        if c not in df.columns:
            raise ValueError(f"column '{c}' not found")

    df = df.copy()
    df[t] = pd.to_numeric(df[t], errors="coerce")
    ev = pd.to_numeric(df[e], errors="coerce")
    if not ev.dropna().isin([0, 1]).all():
        raise ValueError(f"event column '{e}' must be coded 1 = event, 0 = censored")
    df[e] = ev
    n0 = len(df)
    df = df.dropna(subset=[t, e] + covs)
    df = df[df[t] >= 0].reset_index(drop=True)
    clean_steps = (f"Dropped {n0 - len(df)} row(s) with missing/invalid values ({n0} -> {len(df)} rows)."
                   if len(df) < n0 else "No cleaning applied; input used as-is.")

    # Build a numeric design matrix; one-hot encode categoricals (drop first = reference).
    design = df[[t, e]].copy()
    for c in covs:
        s = df[c]
        if s.dtype == object or str(s.dtype).startswith("category") or not np.issubdtype(s.dtype, np.number):
            cats = sorted(s.astype(str).unique())
            for lvl in cats[1:]:
                design[f"{c}[{lvl}]"] = (s.astype(str) == lvl).astype(int)
        else:
            design[c] = pd.to_numeric(s, errors="coerce")

    cph = CoxPHFitter()
    cph.fit(design, duration_col=t, event_col=e)
    summ = cph.summary
    terms = list(summ.index)
    hr = summ["exp(coef)"].to_numpy()
    lo = summ["exp(coef) lower 95%"].to_numpy()
    hi = summ["exp(coef) upper 95%"].to_numpy()
    pv = summ["p"].to_numpy()

    ph_p = np.nan
    try:
        from lifelines.statistics import proportional_hazard_test
        ph = proportional_hazard_test(cph, design, time_transform="rank")
        ph_p = float(ph.summary["p"].min())
    except Exception:
        pass

    n = len(terms)
    w, h = 7.4, max(3.6, 0.5 * n + 2.0)
    fig, ax = plt.subplots(figsize=(w, h))
    yy = np.arange(n)[::-1]
    ax.errorbar(hr, yy, xerr=[hr - lo, hi - hr], fmt="s", color="black",
                capsize=3, ms=6, lw=1.0)
    ax.axvline(1.0, ls=":", color="grey")
    ax.set_xscale("log")
    ax.set_yticks(yy)
    ax.set_yticklabels(terms)
    ax.set_ylim(-0.6, n - 0.4)
    ax.set_xlabel("Hazard ratio (95% CI)")
    for y_, h_, l_, u_, p_ in zip(yy, hr, lo, hi, pv):
        star = " *" if p_ < 0.05 else ""
        ax.annotate(f"{h_:.2f} ({l_:.2f}–{u_:.2f})   P {format_p(p_) if p_ < 0.001 else '= ' + format(p_, '.2g')}{star}",
                    xy=(1.02, y_), xycoords=("axes fraction", "data"), va="center", fontsize=8)
    ax.set_title("Hazard ratio", fontweight="bold")
    apply_pub_style(ax)
    fig.tight_layout()

    stats_df = pd.DataFrame({"term": terms, "HR": hr, "ci_low": lo, "ci_high": hi,
                             "z": summ["z"].to_numpy(), "p_value": pv})
    test_meta = {
        "name": "Cox proportional-hazards model",
        "n": int(len(df)), "events": int(df[e].sum()), "covariates": covs,
        "proportional_hazards": {"test": "proportional_hazard_test (min p)", "p_value": ph_p},
        "hazard_ratios": [{"term": tt, "HR": float(h_), "ci_low": float(l_),
                           "ci_high": float(u_), "p_value": float(p_)}
                          for tt, h_, l_, u_, p_ in zip(terms, hr, lo, hi, pv)],
    }
    methods = _methods(stats_df, ph_p, covs, len(df), int(df[e].sum()))
    return {
        "fig": fig, "stats": stats_df, "test_meta": test_meta, "methods": methods,
        "df_used": df, "clean_steps": clean_steps, "label": "hazard_ratios",
        "width": w, "height": h,
        "build_script": lambda in_name, fig_stub: _script(spec, t, e, covs, in_name, fig_stub),
    }


def _methods(stats_df, ph_p, covs, n, events):
    from importlib.metadata import version as v
    import platform
    sig = stats_df[stats_df["p_value"] < 0.05]
    if len(sig):
        parts = "; ".join(f"{r.term} (HR = {r.HR:.2f}, 95% CI {r.ci_low:.2f}-{r.ci_high:.2f}, P = {format_p(r.p_value)})"
                          for r in sig.itertuples())
        sig_txt = f" Significant predictors: {parts}."
    else:
        sig_txt = " No individual predictor reached significance."
    ph_txt = "" if np.isnan(ph_p) else (
        f" The proportional-hazards assumption was {'violated' if ph_p < 0.05 else 'not violated'}"
        f" (minimum covariate cox.zph P = {format_p(ph_p)}).")
    return (
        f"Hazard ratios were estimated with a Cox proportional-hazards model ({n} subjects, {events} events) "
        f"including {', '.join(covs)}." + sig_txt + ph_txt +
        f" Analyses were performed in Python {platform.python_version()} with lifelines {v('lifelines')}."
    )


def _script(spec, t, e, covs, in_name, fig_stub):
    return f'''#!/usr/bin/env python3
# Standalone reproduction. Run inside the pinned container.
import matplotlib; matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np, pandas as pd
from lifelines import CoxPHFitter

df = pd.read_csv("{in_name}")
covs = {covs}
design = df[["{t}", "{e}"]].copy()
for c in covs:
    s = df[c]
    if s.dtype == object or not np.issubdtype(s.dtype, np.number):
        for lvl in sorted(s.astype(str).unique())[1:]:
            design["%s[%s]" % (c, lvl)] = (s.astype(str) == lvl).astype(int)
    else:
        design[c] = pd.to_numeric(s, errors="coerce")
cph = CoxPHFitter().fit(design, "{t}", "{e}")
s = cph.summary
terms = list(s.index); hr = s["exp(coef)"].to_numpy()
lo = s["exp(coef) lower 95%"].to_numpy(); hi = s["exp(coef) upper 95%"].to_numpy(); pv = s["p"].to_numpy()
n = len(terms); yy = np.arange(n)[::-1]
fig, ax = plt.subplots(figsize=(7.4, max(3.6, 0.5 * n + 2.0)))
ax.errorbar(hr, yy, xerr=[hr - lo, hi - hr], fmt="s", color="black", capsize=3, ms=6, lw=1.0)
ax.axvline(1.0, ls=":", color="grey"); ax.set_xscale("log")
ax.set_yticks(yy); ax.set_yticklabels(terms); ax.set_ylim(-0.6, n - 0.4)
ax.set_xlabel("Hazard ratio (95% CI)"); ax.set_title("Hazard ratio", fontweight="bold")
for y_, h_, l_, u_, p_ in zip(yy, hr, lo, hi, pv):
    ax.annotate("%.2f (%.2f-%.2f)  P = %.2g" % (h_, l_, u_, p_), xy=(1.02, y_), xycoords=("axes fraction", "data"), va="center", fontsize=8)
for sp in ("top", "right"): ax.spines[sp].set_visible(False)
fig.tight_layout(); fig.savefig("{fig_stub}.pdf")
print("Reproduced figure.")
'''
